### ISSUE TITLE
システム構成の設計

### ISSUE THREAD

**Issue by nakahashi** [nakahashi/koko.doc#2]
配置図を書いてシステム構成を設計します。
↓こんな構成にする予定です。
- デプロイ先のPaas : Heroku
- Webフレームワーク : Express, MongoDB, React
- 依存するWebサービス : Twitter, ドコモ雑談対話API

**Comment by iKazuyoshi** [comment]
ごめんすでにチケットあったね…
＃３は後でクローズします。

**Comment by iKazuyoshi** [comment]
>  ドコモ雑談対話API

なにこれ…？
まぁ良いけど、Heroku選定の理由はなんでしょう…（Heorkuってただのタイポだよね？）

**Comment by nakahashi** [comment]
タイポ直しました。
ドコモ雑談対話API → https://dev.smt.docomo.ne.jp/?p=docs.api.page&api_name=dialogue&p_name=api_reference
このAPIで会話できるTwitterボットにするつもりです。

**Comment by nakahashi** [comment]
Herokuを選んだのは、
- 情報が割と多い
- Githubと連携すればデプロイ簡単
- 無料

といったの理由です。
いい候補がほかにもあれば教えてください。

**Comment by nakahashi** [comment]
というか、実はheroku以外知りません。。

**Comment by nakahashi** [comment]
システム構成はこのように考えています。
![layout](https://cloud.githubusercontent.com/assets/13581101/10073584/f944a49c-6304-11e5-9faa-0f9e4b197651.png)
- Node/Express4で作るWebアプリとして、Twitterボットを作ります。
- ボットの設定用WebサイトのViewレイヤライブラリにReactを使いたいです。（強い必然性はないですが、仮想DOMという考え方がこれから息の長い技術になりそうなので、とりあえず触ってみたいです）
- 最初はHerokuで動かしたいですが、落ち着いたら他のレンタルサーバやRaspberry Piでの自家サーバ構築などもやってみたい気もします。

**Comment by nakahashi** [comment]
Ｗｅｂ開発ではタスクランナーとかいうのを使うのが一般的なようなので、gulpというタスクランナーを使ってPlantUMLの`*.pu`ファイルを`*.png`に変換してくれるタスクを勉強用に書いてみました。

PlantUML用のgulpプラグインで文字化けしたので、[文字コードを指定できる機能を追加するPull Request](https://github.com/y-matsuwitter/gulp-plantuml/pull/2)を出してみました。

そういう状態なので、このリポジトリ上のgulpfileはまだ仮です。

**Comment by iKazuyoshi** [comment]
最近はたいていのメジャーなPaas事業者はGithubと連携しているし
仮に中橋リーダーがお客さんなり、うちのお偉いさんとかを
説得する材料とするには弱いかな…

他のサービスと比較してHerokuが良い！と自信を持って言える長所を
いくつかのサービスのPros / Consを洗い出して言えるようになるとベストです。

今回はHerokuで良いけど、実際の業務とかだとこういった点を突っ込まれるので
余裕をもってクリアできるようにしましょう！

**Comment by iKazuyoshi** [comment]
ついでに言うと、私もHerokuは使ったことないので楽しみです。

**Comment by iKazuyoshi** [comment]
Expressを選定した理由は？
http://qiita.com/oukayuka/items/14bfdcb6b5411a2b4b7c
この編を見ると、NodeのWebアプリケーションFWって、人気がばらけてるよね。
Expressは、ちょっとリッチな感じですが、何か選定の理由ってありますか？

**Comment by nakahashi** [comment]
Expressは、Node向けWebフレームワークでは一番初めに紹介されている印象があったので、あまり深く考えずに選びました。実際稲富さんが上げた記事の中でも、Stack OverflowとQiitaでの情報が多かったのは印象と一致しています。
なので、他のフレームワークと比較して選んだわけではありません。

**Comment by nakahashi** [comment]
いや、実はHubotというボット用フレームワークを一度試してみたのですが、Twitterとの接続がいまいちだったので、やめたのでした。
ただ、Hubotも内部ではExpress3を使っていましたので、Expressを使うのは見当違いではないのかと。

**Comment by nakahashi** [comment]
私はWebフレームワークというとRubyのRailsとSinatraしか知らないのですが、ExpressはSinatraに似て比較的シンプルなフレームワークという感じでしたが。。

**Comment by nakahashi** [comment]
ただ、Express以外のフレームワークも少し調べたら面白そうだったので、もう少し調査してみたいと思います。

**Comment by nakahashi** [comment]
調査の途中ですが、少しExpressで作ったボットに関する記事を、Qiitaにアップしました。
[Node.js(Express)で作ったTwitterボットとダイレクトメッセージで雑談する - Qiita](http://qiita.com/nakahashi/items/9566f9cd31e0160aeac2)

**Comment by iKazuyoshi** [comment]
Expressの件はもう少しExpressになにを期待しているのか考えてみましょうか。
ここ、超重要で一度決めると後々変更し辛いのでね・・・。
本当はF2Fで話ししたいとこだね。。。

**Comment by nakahashi** [comment]
「何を期待しているか」という話ですが、それは「他のフレームワークと比べて」、ということでしょうか。それとも、「素のNodeと比べて」ということでしょうか。
前者は調査中です。
後者は以下のような感じです。
- ルーティング
- テンプレートエンジンを使ったレンダリング
- ログ
- エラー処理
- ロジックのモジュール化（middleware化）

**Comment by nakahashi** [comment]
Node用フレームワークを調査して以下にまとめました。
[Node.js - メジャーなNode用Webアプリケーションフレームワークを比較する - Qiita](http://qiita.com/nakahashi/items/15c016ff8beacd014fe4)

Expressもいいですが、非同期処理を支援してくれるKoaがいいかも、と思いました。
Javascript書くとコールバックだらけになるので、そこをスマートに書けるようになるというのはJavascriptプログラマとしてこれから重要になっていくのではないかと。Koaはそこに切り込んでいっているので。

Koaができることとしては、非同期処理の支援を除けばExpressと似たようなものっぽいです。

**Comment by iKazuyoshi** [comment]
よく分からないけど、Githubからアカウントロックされてました...
Heroic、Express使ってみたけど、Expressは良さげですがHerokuとMongoは今後のプロダクト
開発においては恐らく商用利用が安定性とかの観点で難しそうなので、AWS使うようにしませんか？

そうなると、Expressなんかも使わなくなりそうなんですが…

**Comment by nakahashi** [comment]
AWS全然わからないので、教えてください。。
- 結構お金かかっちゃいません？
- ボットの動作設定Webアプリの画面仕様書を作ったのですが、これをフレームワークなし＆AWSでいけるでしょうか？→ [画面仕様書](https://github.com/nakahashi/koko.doc/blob/master/koko-console%E7%94%BB%E9%9D%A2%E4%BB%95%E6%A7%98%E6%9B%B8.md)
- 冒頭のほうに書いた通り勉強のためにViewのフレームワークにReactを使いたいのですが、それもAWSで行けます？

**Comment by nakahashi** [comment]
AWSの情報を少し読んで少し自己解決しました。
こんな構成でどうでしょう？

![aws](https://cloud.githubusercontent.com/assets/13581101/10666935/88d5f8de-790f-11e5-84b6-4691832e95ef.png)

無料枠は各サービスごとにあるように読めるので、たぶん枠内で行けると思うんですけど。

**Comment by nakahashi** [comment]
うーむ、上記の構成だとkokoコーンソールのjavascriptのコードにtwitterやAWSのトークンを直書きしなきゃいけなくてやっぱり駄目ですね。。
サーバーサイドにTwitterやAWSにアクセスするAPIを作りたいのですが、koaは使っちゃダメですか？

**Comment by nakahashi** [comment]
AWS Elastic Beanstalkにkokoボットをデプロイしてみました。
[ボットのID](https://twitter.com/koko_bot15)
[ソースのリポジトリ](https://github.com/nakahashi/koko-bot)

ボットの機能的には以下の記事の時と変わりませんが、ES6で書き直したりしてます。
[Node.js(Express)で作ったTwitterボットとダイレクトメッセージで雑談する - Qiita](http://qiita.com/nakahashi/items/9566f9cd31e0160aeac2)

リポジトリに今後の作業をIssueで挙げています。
kokoコンソールのWebアプリに関しては、フロントエンドのみちょっと作ってみます。
